File: lib/src/main/java/com/walmartlabs/components/scheduler/core/BucketsLoader.java
package com.walmartlabs.components.scheduler.core;

import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.ListenableScheduledFuture;
import com.walmart.gmp.ingestion.platform.framework.shutdown.SupportsShutdown;
import com.walmart.gmp.ingestion.platform.framework.data.core.DataManager;
import com.walmart.gmp.ingestion.platform.framework.data.core.Selector;
import com.walmart.gmp.ingestion.platform.framework.data.core.TaskExecutor;
import com.walmartlabs.components.scheduler.entities.Bucket;
import org.apache.log4j.Logger;

import java.time.ZonedDateTime;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;
import java.util.function.Predicate;

import static com.google.common.util.concurrent.Futures.addCallback;
import static com.walmart.gmp.ingestion.platform.framework.core.Props.PROPS;
import static com.walmart.gmp.ingestion.platform.framework.shutdown.ShutdownRegistry.SHUTDOWN_REGISTRY;
import static com.walmart.gmp.ingestion.platform.framework.data.core.DataManager.retryableExceptions;
import static com.walmart.gmp.ingestion.platform.framework.data.core.Selector.fullSelector;
import static com.walmart.platform.soa.common.exception.util.ExceptionUtil.getRootCause;
import static com.walmartlabs.components.scheduler.core.BucketManager.createEmptyBucket;
import static com.walmartlabs.components.scheduler.core.ScheduleScanner.SCHEDULER;
import static java.lang.String.format;
import static java.time.ZonedDateTime.now;
import static java.util.concurrent.TimeUnit.SECONDS;

/**
 * Created by smalik3 on 7/5/16
 */
class BucketsLoader implements SupportsShutdown {

    private static final Logger L = Logger.getLogger(BucketsLoader.class);

    private final int lookbackRange;
    private final int fetchSize;
    private final Consumer<Bucket> consumer;
    private final Predicate<ZonedDateTime> predicate;
    private final int bucketWidth;
    private final DataManager<ZonedDateTime, Bucket> dataManager;
    private final ZonedDateTime bucketId;
    private final int waitInterval;
    private final TaskExecutor taskExecutor = new TaskExecutor(retryableExceptions);
    private static final Selector<ZonedDateTime, Bucket> SELECTOR = fullSelector(now());
    private final AtomicReference<ListenableScheduledFuture<?>> runningJob = new AtomicReference<>();

    BucketsLoader(int lookbackRange, int fetchSize, Consumer<Bucket> consumer,
                  Predicate<ZonedDateTime> predicate, int bucketWidth,
                  DataManager<ZonedDateTime, Bucket> dataManager, ZonedDateTime bucketId) {
        this.lookbackRange = lookbackRange;
        this.fetchSize = fetchSize;
        this.consumer = consumer;
        this.predicate = predicate;
        this.bucketWidth = bucketWidth;
        this.dataManager = dataManager;
        this.bucketId = bucketId;
        waitInterval = PROPS.getInteger("buckets.background.load.wait.interval.seconds", 15);
        SHUTDOWN_REGISTRY.register(this);
    }

    void start() {
        L.info(format("starting the background load of buckets at a rate of %d buckets per %d seconds until %d buckets are loaded", fetchSize, waitInterval, lookbackRange));
        runningJob.set(SCHEDULER.schedule(() -> load(0), 0, SECONDS));
    }

    private void load(int fromIndex) {
        if (fromIndex >= lookbackRange) {
            L.info("lookback range reached, no more buckets will be loaded in background");
            return;
        }
        try {
            L.info("initiating background load of buckets from index: " + fromIndex);
            final AtomicReference<Integer> currentBucketIndex = new AtomicReference<>();
            final AtomicBoolean atLeastOne = new AtomicBoolean();
            for (int i = 1; i <= fetchSize; i++) {
                final int bucketIndex = fromIndex + i;
                if (bucketIndex <= lookbackRange) {
                    currentBucketIndex.set(bucketIndex);
                    final ZonedDateTime bId = bucketId.minusSeconds(bucketIndex * bucketWidth);
                    if (!predicate.test(bId)) {
                        atLeastOne.set(true);
                        L.info(format("loading bucket: %s, failures will be retried %d times every %d seconds", bId, lookbackRange - bucketIndex + 1, bucketWidth));
                        addCallback(taskExecutor.async(() -> dataManager.getAsync(bId, SELECTOR), "bucket-load:" + bId,
                                lookbackRange - bucketIndex + 1, bucketWidth, 1, SECONDS), new FutureCallback<Bucket>() {
                            @Override
                            public void onSuccess(Bucket b) {
                                L.info(format("bucket %s loaded successfully", bId));
                                consumer.accept(b == null ? createEmptyBucket(bId) : b);
                            }

                            @Override
                            public void onFailure(Throwable t) {
                                L.error(format("error in loading bucket %s, system is giving up", bId), getRootCause(t));
                            }
                        });
                    } else {
                        L.info(format("bucket %s already loaded, skipping...", bId));
                    }
                } else {
                    L.info("no more buckets to load, look back range reached");
                    break;
                }
            }
            runningJob.set(SCHEDULER.schedule(() -> load(currentBucketIndex.get()), !atLeastOne.get() ? 0 : waitInterval, SECONDS));
        } catch (Exception e) {
            L.error("could not load buckets: " + this, e);
        }
    }

    @Override
    public String name() {
        return "BucketLoader";
    }

    @Override
    public int priority() {
        return 0;
    }

    @Override
    public ListenableFuture<?> shutdown() {
        if (runningJob.get() != null)
            runningJob.get().cancel(true);
        return runningJob.get();
    }

    @Override
    public String toString() {
        return "BucketsLoader{" +
                "lookbackRange=" + lookbackRange +
                ", fetchSize=" + fetchSize +
                ", consumer=" + consumer +
                ", predicate=" + predicate +
                ", bucketWidth=" + bucketWidth +
                ", dataManager=" + dataManager +
                ", bucketId=" + bucketId +
                '}';
    }
}

File: lib/src/main/kotlin/com/walmartlabs/bigben/core/BucketsLoader.kt
/*-
 * #%L
 * BigBen:lib
 * =======================================
 * Copyright (C) 2016 - 2018 Walmart Inc.
 * =======================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package com.walmartlabs.bigben.core

import com.google.common.util.concurrent.ListenableScheduledFuture
import com.walmartlabs.bigben.core.BucketManager.Companion.scheduler
import com.walmartlabs.bigben.entities.Bucket
import com.walmartlabs.bigben.extns.fetch
import com.walmartlabs.bigben.utils.done
import com.walmartlabs.bigben.utils.logger
import com.walmartlabs.bigben.utils.rootCause
import com.walmartlabs.bigben.utils.utils.Props
import com.walmartlabs.bigben.utils.utils.TaskExecutor
import java.time.ZonedDateTime
import java.util.concurrent.TimeUnit.SECONDS
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Predicate

/**
 * Created by smalik3 on 2/22/18
 */
class BucketsLoader(private val lookbackRange: Int, private val fetchSize: Int, private val predicate: Predicate<ZonedDateTime>,
                    private val bucketWidth: Int, private val bucketId: ZonedDateTime, private val consumer: (Bucket) -> Unit) : Runnable {

    companion object {
        private val l = logger<BucketsLoader>()
    }

    private val waitInterval = Props.int("buckets.background.load.wait.interval.seconds")
    private val runningJob = AtomicReference<ListenableScheduledFuture<*>>()
    private val taskExecutor = TaskExecutor(setOf(Exception::class.java))

    override fun run() {
        l.info("starting the background load of buckets at a rate of {} buckets per {} seconds until {} buckets are loaded", fetchSize, waitInterval, lookbackRange)
        runningJob.set(scheduler.schedule({ load(0) }, 0, SECONDS))
    }

    private fun load(fromIndex: Int) {
        if (fromIndex >= lookbackRange) {
            if (l.isInfoEnabled) l.info("lookback range reached, no more buckets will be loaded in background")
        } else {
            if (l.isInfoEnabled) l.info("initiating background load of buckets from index: {}", fromIndex)
            val currentBucketIndex = AtomicReference<Int>()
            val atLeastOne = AtomicBoolean()
            (1..fetchSize).forEach {
                val bucketIndex = fromIndex + it
                if (bucketIndex <= lookbackRange) {
                    currentBucketIndex.set(bucketIndex)
                    val bId = bucketId.minusSeconds((bucketIndex * bucketWidth).toLong())
                    if (!predicate.test(bId)) {
                        atLeastOne.set(true)
                        if(l.isDebugEnabled) l.debug("loading bucket: {}, failures will be retried {} times, every {} seconds", bId, lookbackRange - bucketIndex + 1, bucketWidth)
                        taskExecutor.async("bucket-load:$bId", lookbackRange - bucketIndex + 1, bucketWidth, 1) { fetch<Bucket> { it.bucketId = bId } }
                                .done({ l.error("error in loading bucket {}, system is giving up", bId, it.rootCause()) }) {
                                    if (l.isDebugEnabled) l.debug("bucket {} loaded successfully", bId)
                                    consumer(it ?: BucketManager.emptyBucket(bId))
                                }
                    } else {
                        if (l.isDebugEnabled) l.debug("bucket {} already loaded (likely by checkpoint), skipping...", bId)
                    }
                }
            }
            runningJob.set(scheduler.schedule({ load(currentBucketIndex.get()) }, (if (!atLeastOne.get()) 0 else waitInterval).toLong(), SECONDS))
        }
    }
}
